### PR TITLE
Remote config (I): create infrastructure

### PR DESCRIPTION
**Pull request by marcosholgado**
Task/Issue URL: https://app.asana.com/0/1125189844152671/1200676769415755
Tech Design URL: https://app.asana.com/0/1125189844152671/1200721087084600
CC: 

**Description**:
This PR adds the basic infrastructure to be able to use remote config. For now it downloads the privacy config locally and stores some of the data.

**Steps to test this PR**:
1. Install and launch the app
1. In the logcat you should see `Downloading privacy config`
1. Check the database of the privacy config module
1. The table `content_blocking_exceptions` should match the exceptions lists from the contentBlocking section in `https://staticcdn.duckduckgo.com/trackerblocking/config/v1/android-config.json`
1. The table `toggles` should have `contentBlocking` and enabled as 1.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyConfig.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+package com.duckduckgo.privacy.config.api
+
+import com.duckduckgo.feature.toggles.api.FeatureName
+
+interface PrivacyConfigDownloader {
+    suspend fun download(): Boolean
+}
```
Comment (aitorvs) [inline privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyConfig.kt:23]: This seems to only be used by the `privacy-config-impl`. 
Do you think the `download` method will be necessary for any other module that it's not `privacy-config-impl`. If only the `impl` is going to use it, maybe we can move this interface to the `impl` for now?

File: feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.feature.toggles.api
+
+interface FeatureTogglesPlugin {
```
Comment (aitorvs) [inline feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt:19]: Adding some javadocs as it is public API
```suggestion
/**
 * Features that can be enabled/disabled should implement this plugin.
 * The associated plugin point will call the plugins when when the [FeatureToggle] API
 * is used
 */
interface FeatureTogglesPlugin {
```

File: feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    fun isEnabled(featureName: FeatureName, defaultValue: Boolean): Boolean?
```
Comment (aitorvs) [inline feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt:20]: Adding some javadocs as it is public API
```suggestion
    /**
     * This method will return whether the plugin knows about the [featureName] in which case will return
     * whether it is enabled or disabled
     * @return `true` if the feature is enable. `false` when disabled. `null` if the plugin does not know the featureName
     */
    fun isEnabled(featureName: FeatureName, defaultValue: Boolean): Boolean?
```

File: feature-toggles/feature-toggles-impl/src/main/java/com/duckduckgo/feature/toggles/impl/RealFeatureTogglesImpl.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import com.duckduckgo.app.global.plugins.PluginPoint
+import com.duckduckgo.di.scopes.AppObjectGraph
+import com.duckduckgo.feature.toggles.api.FeatureToggle
+import com.duckduckgo.feature.toggles.api.FeatureTogglesPlugin
+import com.duckduckgo.feature.toggles.api.FeatureName
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+@ContributesBinding(AppObjectGraph::class)
+class RealFeatureToggleImpl @Inject constructor(private val featureTogglesPluginPoint: PluginPoint<FeatureTogglesPlugin>) :
```
Comment (aitorvs) [inline feature-toggles/feature-toggles-impl/src/main/java/com/duckduckgo/feature/toggles/impl/RealFeatureTogglesImpl.kt:28]: I think we may want the instance to be a singleton
```suggestion
import javax.inject.Inject
import javax.inject.Singleton

@ContributesBinding(AppObjectGraph::class)
@Singleton
class RealFeatureToggleImpl @Inject constructor(private val featureTogglesPluginPoint: PluginPoint<FeatureTogglesPlugin>) :
```

File: privacy-config/privacy-config-store/src/main/java/com/duckduckgo/privacy/config/store/PrivacyConfigDatabase.kt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    abstract fun privacyFeatureTogglesDao(): PrivacyFeatureTogglesDao
+
+    companion object {
+        @Volatile
+        private var INSTANCE: PrivacyConfigDatabase? = null
+
+        fun getInstance(context: Context): PrivacyConfigDatabase =
+            INSTANCE ?: synchronized(this) {
+                INSTANCE ?: buildDatabase(context).also { INSTANCE = it }
+            }
```
Comment (aitorvs) [inline privacy-config/privacy-config-store/src/main/java/com/duckduckgo/privacy/config/store/PrivacyConfigDatabase.kt:43]: I wonder if, given that the `providePrivacyConfigDatabase` is marked as `Singleton` already, we can probably remove all the ceremony to create a singleton manually (?)

**Review by aitorvs — APPROVED** [review]
This is extremely nice Marcos. 
Left couple suggestions. I think the only one that should really change is annotating `RealFeatureToggleImpl` as `@Singleton`

File: privacy-config/privacy-config-store/src/main/java/com/duckduckgo/privacy/config/store/PrivacyConfigDatabase.kt (hunk 5 above)
Comment (marcosholgado) [reply privacy-config/privacy-config-store/src/main/java/com/duckduckgo/privacy/config/store/PrivacyConfigDatabase.kt]: Makes sense, changed!

File: feature-toggles/feature-toggles-impl/src/main/java/com/duckduckgo/feature/toggles/impl/RealFeatureTogglesImpl.kt (hunk 4 above)
Comment (marcosholgado) [reply feature-toggles/feature-toggles-impl/src/main/java/com/duckduckgo/feature/toggles/impl/RealFeatureTogglesImpl.kt]: Yup, done!

File: feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt (hunk 3 above)
Comment (marcosholgado) [reply feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt:31]: 👍

File: feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt (hunk 2 above)
Comment (marcosholgado) [reply feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureTogglesPlugin.kt:24]: 👍

File: privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyConfig.kt (hunk 1 above)
Comment (marcosholgado) [reply privacy-config/privacy-config-api/src/main/java/com/duckduckgo/privacy/config/api/PrivacyConfig.kt]: It made sense when download was managed by the features module but not anymore. Moved to `impl`!

**Pull request merged**